File: javascript/20.valid-parentheses.js
/*
 * @lc app=leetcode id=20 lang=javascript
 *
 * [20] Valid Parentheses
 */

// @lc code=start
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
  if (s.length === 0 || s.length % 2 !== 0) return false
  let stack = []
  for (let c of s) {
    let top = stack[stack.length - 1]
    if (c === "}" && top === "{") {
      stack.pop()
      continue
    }
    if (c === ")" && top === "(") {
      stack.pop()
      continue
    }
    if (c === "]" && top === "[") {
      stack.pop()
      continue
    }
    stack.push(c)
  }
  return stack.length === 0
};
// @lc code=end

File: cpp/20.valid-parentheses.cpp
/*
 * @lc app=leetcode id=20 lang=cpp
 *
 * [20] Valid Parentheses
 */

// @lc code=start
class Solution
{
public:
    bool isValid(string s)
    {
        stack<char> stacker;
        for (char &c : s)
        {
            switch (c)
            {
            case '(':
            case '[':
            case '{':
                stacker.push(c);
                break;
            case ')':
                if (stacker.empty())
                {
                    return false;
                }
                if (stacker.top() == '(')
                {
                    stacker.pop();
                }
                else
                {
                    stacker.push(c);
                }
                break;
            case ']':
                if (stacker.empty())
                {
                    return false;
                }
                if (stacker.top() == '[')
                {
                    stacker.pop();
                }
                else
                {
                    stacker.push(c);
                }
                break;
            case '}':
                if (stacker.empty())
                {
                    return false;
                }
                if (stacker.top() == '{')
                {
                    stacker.pop();
                }
                else
                {
                    stacker.push(c);
                }
                break;
            }
        }
        return stacker.empty();
    }
};
// @lc code=end
